hazelcast/src/main/java/com/hazelcast/ringbuffer/impl/ArrayRingbuffer.java

```diff
@@ -135,4 +144,41 @@ private int toIndex(long sequence) {
     public void set(long seq, T data) {
         ringItems[toIndex(seq)] = data;
     }
+
+    @Override
+    public void setSerializationService(SerializationService serializationService) {
+        this.serializationService = serializationService;
+    }
+
+    @Override
+    public Long merge(SplitBrainMergeEntryView<Long, T> mergingEntry, SplitBrainMergePolicy mergePolicy) {
+        mergePolicy.setSerializationService(serializationService);
+
+        // try to find an existing item with the same value
+        T existingItem = null;
+        long existingSequence = -1;
+        for (long sequence = headSequence; sequence <= tailSequence; sequence++) {
+            T item = read(sequence);
+            if (mergingEntry.getValue().equals(item)) {
```

Comment:
The type of the items in the array could be `Data` if the ringbuffer in-memory-format is `BINARY` so this check will not pass. Also, I think we should be merging entries with the same sequence, not the same values. In an extreme scenario, the ringbuffer could contain all of the same items. A sequence is an identifier for a ringbuffer entry so if the item exists on both clusters with the same sequence, those should be merged.

Reply:
The check works fine:
![screenshot from 2017-12-18 12-51-22](https://user-images.githubusercontent.com/4196298/34104800-3068b768-e3f2-11e7-8704-ec51fb5b71a4.png)

Reply:
Are you testing both with `InMemoryFormat.BINARY` and `InMemoryFormat.OBJECT` for the ringbuffer in memory format?

Reply:
At least for the "custom" merge policy, yes.
```
        return asList(new Object[][]{
                {BINARY, DiscardMergePolicy.class},
                {BINARY, PassThroughMergePolicy.class},
                {BINARY, PutIfAbsentMergePolicy.class},

                {BINARY, MergeIntegerValuesMergePolicy.class},
                {OBJECT, MergeIntegerValuesMergePolicy.class},
        });
```
But I also get the coverage when I run with `OBJECT` and `PutIfAbsentMergePolicy`. I guess it's clearer now, since the type is `<E>`, not the duplicated `<T>` anymore.